cmd/snap-bootstrap/partition/encrypt.go

```diff
@@ -133,3 +161,61 @@ func cryptsetupClose(name string) error {
 	}
 	return nil
 }
+
+func cryptsetupAddKey(key EncryptionKey, rkey RecoveryKey, node string) error {
+	// create a named pipe to pass the recovery key
+	fpath := filepath.Join(dirs.SnapRunDir, "tmp-rkey")
+	if err := syscall.Mkfifo(fpath, 0600); err != nil {
+		return fmt.Errorf("cannot create named pipe: %v", err)
+	}
+	defer os.RemoveAll(fpath)
+
+	// add a new key to slot 1 reading the passphrase from the named pipe
+	args := []string{
+		// add a new key
+		"luksAddKey",
+		// the encrypted device
+		node,
+		// batch processing, no password verification
+		"-q",
+		// read existing key from stdin
+		"--key-file", "-",
+		// store it in keyslot 1
+		"--key-slot", "1",
+		// the named pipe
```

Comment:
Reading the manual page for `cryptsetup` and specifically the section on `luksAddKey` I cannot find any reference to further arguments. I understand that the named pipe is written to by `cryptsetup` and we write the recovery key to it but I cannot find any documentation for it.

Reply:
The documentation mentions that "the new passphrase to be added can be specified interactively or read from the file given as positional argument". We're using the named pipe as the positional argument to specify the new passphrase.

Reply:
So stdin is one thing and named pipe via `fpath` below is something else. Or am I confused and just misread this?

Reply:
You're right, they are two different things. You need to supply the original key and the new key. The former goes through stdin and the latter using the named pipe, to avoid creating any temporary file.